keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,925 @@
+
+## Constraints and Conflict Resolution
+
+Exported services are derived from the properties of each component service and
+their respective endpoints. However, some properties combine across exports
+better than others. They generally fall into two categories: global properties,
+and component-level properties.
+
+
+### Global Properties
+
+These properties describe how the service should be consumed as a whole. They
+directly impact service consumption and must be consistent across all child
+services. If these properties are out of sync for a subset of exported services,
+there is no clear way to determine how a service should be accessed. **If any
+global properties have conflicts that can not be resolved, a condition will be
+set on the `ServiceExport` with a description of the conflict. The service will
+not be synced, and an error will be set on the status of each affected
+`ServiceExport` and any previously-derived `ServiceImports` will be deleted
+from each cluster in the supercluster.**
+
+
+#### Service Port
+
+A derived service will be accessible with the supercluster IP at the ports
+dictated by child services. If the external properties of service ports for a
+set of exported services don’t match, we won’t know which port is the correct
+choice for a service. For example, if two exported services use different ports
+with the name “http”, which port is correct? What if a service uses the same
+port with different names? As long as there are no conflicts (different ports
+with the same name), the supercluster service will expose the superset of
+service ports declared on its component services. If a user wants to change a
+service port in a conflicting way, we recommend deploying a new service or
+making the change in non-conflicting phases.
+
+
+#### IP Family
+
+Because IPv4 and IPv6 addresses cannot be safely intermingled (e.g. iptables
+rules can not mix IPv4 and IPv6), all component exported services making up a
+supercluster service must use the same `IPFamily`. As with the single cluster
+case - a service’s `IPFamily` is immutable - changing families will require a
+new service to be created.
+
+
+### Component Level Properties
+
+These properties are export-specific and pertain only to the subset of endpoints
+backed by a single instance of each exported service. They may be safely carried
+throughout the supercluster without risk of conflict. We propagate these
+properties forward with no attempt to merge or alter them.
+
+
+#### Session Affinity
+
+Session affinity affects a service as a whole for a given consumer. What would
+it mean for a service to have e.g. client IP session affinity set for half its
+backends? Would sessions only be sticky for those backends, or would there be no
+affinity? If sessions are selectively sticky, we’d expect to see traffic to skew
+toward the sticky subset of endpoints. That said, there’s nothing preventing us
+from applying affinity on a per-slice basis so we will carry it forward.
```

Comment:
This will impose a constraint on implementations to support this...

Reply:
Yes it will. This behavior seems easiest to reason about, but weighed against other factors maybe we need to take a different approach. Merging here isn't necessarily simple. Example alternatives I see:

- SessionAffinity is None unless all clusters agree
- SessionAffinity uses client IP if any clusters use client IP
- If using client IP, use the minimum timeout value

In any case, the user needs to know about our merge logic and cluster state to reason about what's happening, and it would have global impact.

Reply:
Let's say the VIP was implemented by an off-cluster LB.
I'm not sure the "session affinity" on a subset is something that is possible to do in most LB products.
Do we really need to mandate this?